### ISSUE TITLE
Profile Pose Estimation & Improve Performance Where and As Necessary

### ISSUE THREAD

**Issue by RobertLeahy** [RobertLeahy/SENG499#179]
We're down to <85 ms per frame on pose estimation on my PC.

<130 ms per frame overall on my PC.

**Comment by RobertLeahy** [comment]
Currently the `correspondences` kernel creates a bunch of A matrices and B vectors and stores them in global memory.  We then load all these matrices and vectors from global memory to local memory to perform the parallel tree-based sum.

Why not just find the correspondences, create the A matrix and B vector and rather than storing them in global memory proceed directly to the first iteration of the parallel tree-based sum?

This should save us >8 million global reads and writes per iteration.

**Comment by RobertLeahy** [comment]
Using [`__constant`](https://www.khronos.org/registry/cl/sdk/1.1/docs/man/xhtml/constant.html) instead of `__global` in certain places leads to a ~5 ms per frame speed up with 15 iterations.

See 31eb1c9cda53fdd595976c098ccbd4c155193873.

**Comment by RobertLeahy** [comment]
I've been fooling around with having each work group load T_g,k^-1, T_z, and K into `__local` buffers before proceeding with `correspondences`.  This leads to a ~2 ms speed up (may be as low as ~1 ms).  More investigation of this is necessary due to the fact the speed up seems to vary with the work group size (higher work group sizes (32 vs. 16) seem to give an improvement to a point, but this degrades once they get too large (tried a work group size of 640)).

See a1b876cb4f5b6d7fb14f059b1cff57186ae2df3a.

I've left this on its own branch (`#179-performance-local`) because I'm not convinced that it's worthwhile, and because I'm not convinced that my implementation is a good one (there should be more effort put into choosing work group sizes other than just leaving it to the implementation, perhaps).

**Comment by RobertLeahy** [comment]
Merged the first iteration of the tree-based sum into the `correspondences` kernel.  This allows the `correspondences` kernel to write mostly into local memory with a single write to global memory at the end for each work group.  This massively reduces I/O to global memory which leads to a massive speed up: ~80 ms per frame down to ~20 ms (~4x improvement).

See: 79f7e67938c4e452fbd3be44ca8a4e72f6c4f1fd.

**Comment by RobertLeahy** [comment]
[This guide is intense](http://developer.amd.com/tools-and-sdks/opencl-zone/amd-accelerated-parallel-processing-app-sdk/opencl-optimization-guide/), just going to leave it here for reference.

**Comment by RobertLeahy** [comment]
I tried another alternative to 31eb1c9cda53fdd595976c098ccbd4c155193873: I made structs which contain an array large enough for a 4x4 or 3x3 matrix and the `correspondences` kernel accepts these by value.  My theory was that the `__private` memory (which holds kernel arguments) would be faster than `__constant` memory.

Once again the performance doesn't change much (maybe down by 1 ms but that could easily be noise).  My theory is that these matrices are so hot that they're pretty much constantly in cache somewhere.  The fact that they're `__constant` and `const` and therefore never written makes it even better.

See: 56faa172a1f55b59f85a5d20934f0f0696eaf935.
